optimum/habana/transformers/generation/utils.py

```diff
@@ -2354,6 +2361,12 @@ def _sample(
 
             if token_idx is not None:
                 idx = token_idx + model_kwargs.get("inputs_embeds_offset", 0)
+                if batch_size > 1 and has_eos_stopping_criteria:
+                    for i in range(batch_size):
+                        if next_tokens[i] == eos_token_id:
+                            idx_bs[i] = token_idx.item()
+                        if token_idx > idx_bs[i]:
+                            next_tokens[i] = pad_token_id
```

Comment:
please do not add any logic inside the generation loop, it will impact performance. instead you can do it after generation  to return right content based on eos

Reply:
moved logic after generation loop and before return